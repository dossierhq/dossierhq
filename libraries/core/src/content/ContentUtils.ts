import type {
  AdminEntity,
  AdminEntityCreate,
  EntityLike,
  Location,
  PublishedEntity,
  RichText,
  RichTextNode,
} from '../Types.js';
import { FieldType, type FieldSpecification } from '../schema/SchemaSpecification.js';
import { assertExhaustive } from '../utils/Asserts.js';
import type { Mutable } from '../utils/TypeUtils.js';
import type { ContentValuePath } from './ContentPath.js';
import { isRichTextRootNode } from './ContentTypeUtils.js';

export function getEntityNameBase(name: string): string {
  const hashIndex = name.lastIndexOf('#');
  if (hashIndex < 0) {
    return name;
  }
  return name.slice(0, hashIndex);
}

export function isEntityNameAsRequested(currentName: string, requestedName: string): boolean {
  if (requestedName === currentName) {
    return true;
  }
  const currentNameBase = getEntityNameBase(currentName);
  return requestedName === currentNameBase;
}

export function copyEntity<
  T extends
    | AdminEntity<string, object>
    | AdminEntityCreate<AdminEntity<string, object>>
    | PublishedEntity<string, object>
    | EntityLike<string, object>,
>(
  entity: Readonly<T> | T,
  changes: { id?: string; info?: Partial<T['info']>; fields?: Partial<T['fields']> },
): Readonly<T> {
  const copy: Mutable<T> = { ...entity };
  if (typeof changes.id === 'string') {
    copy.id = changes.id;
  }
  if (changes.info) {
    copy.info = { ...entity.info };
    for (const [key, value] of Object.entries(changes.info)) {
      (copy.info as unknown as Record<string, unknown>)[key] = value;
    }
  }
  if (changes.fields) {
    const fieldsCopy = { ...entity.fields } as Record<string, unknown>;
    copy.fields = fieldsCopy;
    for (const [key, value] of Object.entries(changes.fields)) {
      fieldsCopy[key] = value;
    }
  }
  return copy;
}

export function checkFieldTraversable(
  fieldSpec: FieldSpecification,
  value: unknown,
): { path: ContentValuePath; message: string } | null {
  if (fieldSpec.list) {
    if (value !== null && value !== undefined && !Array.isArray(value)) {
      return { path: [], message: `Expected a list of ${fieldSpec.type}, got ${typeof value}` };
    }
  }
  return null;
}

export function checkFieldItemTraversable(
  fieldSpec: FieldSpecification,
  value: unknown,
): { path: ContentValuePath; message: string } | null {
  const { type } = fieldSpec;

  if (value === null || value === undefined) {
    return null;
  }

  if (Array.isArray(value)) {
    return { path: [], message: `Expected single ${type}, got a list` };
  }

  switch (type) {
    case FieldType.Boolean: {
      if (typeof value !== 'boolean') {
        return { path: [], message: `Expected a boolean, got ${typeof value}` };
      }
      break;
    }
    case FieldType.Component: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a Component object, got ${typeof value}` };
      }
      if (!('type' in value)) {
        return { path: ['type'], message: `Missing a Component type` };
      }
      if (typeof value.type !== 'string') {
        return { path: ['type'], message: `Expected a Component type, got ${typeof value}` };
      }
      break;
    }
    case FieldType.Reference: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected an entity reference, got ${typeof value}` };
      }
      if (typeof (value as EntityLike).id !== 'string') {
        return {
          path: ['id'],
          message: `Expected an entity reference id, got ${typeof (value as EntityLike).id}`,
        };
      }
      break;
    }
    case FieldType.Location: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a Location object, got ${typeof value}` };
      }
      const { lat, lng } = value as Location;
      if (typeof lat !== 'number' || typeof lng !== 'number') {
        return {
          path: [],
          message: `Expected {lat: number, lng: number}, got {lat: ${typeof lat}, lng: ${typeof lng}}`,
        };
      }
      break;
    }
    case FieldType.Number: {
      if (typeof value !== 'number') {
        return { path: [], message: `Expected a number, got ${typeof value}` };
      }
      break;
    }
    case FieldType.RichText: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a RichText object, got ${typeof value}` };
      }
      const root = (value as RichText).root;

      if (!root) {
        return { path: [], message: `RichText object is missing root` };
      }

      const rootTraversalError = checkRichTextNodeTraversable(root);
      if (rootTraversalError) {
        return { path: ['root', ...rootTraversalError.path], message: rootTraversalError.message };
      }

      if (!isRichTextRootNode(root)) {
        return {
          path: ['root'],
          message: `RichText root is not a valid RichText node, (got ${
            (root as RichTextNode).type
          })`,
        };
      }
      break;
    }
    case FieldType.String: {
      if (typeof value !== 'string') {
        return { path: [], message: `Expected a string, got ${typeof value}` };
      }
      break;
    }
    default:
      assertExhaustive(type);
  }
  return null;
}

export function checkRichTextNodeTraversable(
  node: RichTextNode,
): { path: ContentValuePath; message: string } | null {
  if (!node || typeof node !== 'object') {
    return { path: [], message: `Expected a RichText node, got ${typeof node}` };
  }
  if (typeof node.type !== 'string') {
    return { path: [], message: `RichText node is missing type` };
  }
  if ('children' in node && !Array.isArray(node.children)) {
    return { path: [], message: `RichText node children is not an array` };
  }
  return null;
}
