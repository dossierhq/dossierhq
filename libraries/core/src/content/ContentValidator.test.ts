import { describe, expect, test } from 'vitest';
import type { AdminEntity, AdminEntityCreate, EntityLike } from '../Types.js';
import { AdminSchema } from '../schema/AdminSchema.js';
import { FieldType } from '../schema/SchemaSpecification.js';
import {
  normalizeEntityFields,
  type ContentNormalizerEntityFieldsOptions,
} from './ContentNormalizer.js';
import { ContentTraverseNodeErrorType, traverseEntity } from './ContentTraverser.js';
import { copyEntity } from './ContentUtils.js';
import {
  groupValidationIssuesByTopLevelPath,
  validateEntityInfo,
  validateEntityInfoForCreate,
  validateEntityInfoForUpdate,
  validateTraverseNodeForPublish,
  validateTraverseNodeForSave,
  type PublishValidationIssue,
  type SaveValidationIssue,
} from './ContentValidator.js';
import {
  createRichText,
  createRichTextParagraphNode,
  createRichTextTextNode,
} from './RichTextUtils.js';

const adminSchema = AdminSchema.createAndValidate({
  entityTypes: [
    { name: 'ReferencesEntity', fields: [{ name: 'normal', type: FieldType.Reference }] },
    {
      name: 'LocationsEntity',
      fields: [{ name: 'normal', type: FieldType.Location }],
    },
    {
      name: 'NumbersEntity',
      fields: [{ name: 'integer', type: FieldType.Number, integer: true }],
    },
    {
      name: 'RichTextsEntity',
      fields: [{ name: 'anyNodes', type: FieldType.RichText }],
    },
    {
      name: 'StringsEntity',
      authKeyPattern: 'defaultOrSubject',
      fields: [
        { name: 'normal', type: FieldType.String },
        { name: 'required', type: FieldType.String, required: true },
        { name: 'requiredAndAdminOnly', type: FieldType.String, adminOnly: true, required: true },
        { name: 'multiline', type: FieldType.String, multiline: true },
        { name: 'pattern', type: FieldType.String, matchPattern: 'fooBarBaz' },
        { name: 'patternList', type: FieldType.String, list: true, matchPattern: 'fooBarBaz' },
        {
          name: 'values',
          type: FieldType.String,
          values: [{ value: 'foo' }, { value: 'bar' }, { value: 'baz' }],
        },
        {
          name: 'valuesList',
          type: FieldType.String,
          list: true,
          values: [{ value: 'foo' }, { value: 'bar' }, { value: 'baz' }],
        },
      ],
    },
    {
      name: 'ComponentsEntity',
      fields: [
        { name: 'any', type: FieldType.Component },
        {
          name: 'stringsComponent',
          type: FieldType.Component,
          componentTypes: ['StringsComponent'],
        },
      ],
    },
  ],
  componentTypes: [
    { name: 'AdminOnlyComponent', adminOnly: true, fields: [] },
    {
      name: 'NumbersComponent',
      fields: [{ name: 'integer', type: FieldType.Number, integer: true }],
    },
    { name: 'StringsComponent', fields: [{ name: 'string', type: FieldType.String }] },
  ],
  patterns: [
    { name: 'fooBarBaz', pattern: '^(foo|bar|baz)$' },
    { name: 'defaultOrSubject', pattern: '^(|subject)$' },
  ],
}).valueOrThrow();

const ENTITIES_ENTITY_DEFAULT: AdminEntityCreate = {
  info: { type: 'ReferencesEntity', name: 'ReferencesEntity' },
  fields: {},
};

const LOCATIONS_ENTITY_DEFAULT: AdminEntityCreate = {
  info: { type: 'LocationsEntity', name: 'LocationsEntity' },
  fields: {},
};

const NUMBERS_ENTITY_CREATE_DEFAULT: AdminEntityCreate = {
  info: { type: 'NumbersEntity', name: 'NumbersEntity' },
  fields: {},
};

const STRINGS_ENTITY_CREATE_DEFAULT: AdminEntityCreate = {
  info: { type: 'StringsEntity', name: 'StringsEntity' },
  fields: { required: '-' },
};

const STRINGS_ENTITY_DEFAULT: AdminEntity = {
  id: '123',
  info: {
    type: 'StringsEntity',
    name: 'StringsEntity',
    authKey: '',
    version: 1,
    status: 'draft',
    valid: true,
    validPublished: null,
    createdAt: new Date('2023-04-04T19:26:16.546Z'),
    updatedAt: new Date('2023-04-04T19:26:16.546Z'),
  },
  fields: {},
};

const RICH_TEXTS_ENTITY_CREATE_DEFAULT: AdminEntityCreate = {
  info: { type: 'RichTextsEntity', name: 'RichTextsEntity' },
  fields: {},
};

const VALUE_ITEMS_ENTITY_CREATE_DEFAULT: AdminEntityCreate = {
  info: { type: 'ComponentsEntity', name: 'ComponentsEntity' },
  fields: {},
};

function validateEntity(entity: EntityLike, options?: ContentNormalizerEntityFieldsOptions) {
  const normalizedEntity = {
    ...entity,
    fields: normalizeEntityFields(adminSchema, ['entity'], entity, options).valueOrThrow(),
  };

  const errors: (SaveValidationIssue | PublishValidationIssue)[] = [];
  for (const node of traverseEntity(adminSchema, ['entity'], normalizedEntity)) {
    const error = validateTraverseNodeForSave(adminSchema, node);
    if (error) {
      errors.push(error);
    }
  }

  for (const node of traverseEntity(
    adminSchema.toPublishedSchema(),
    ['entity'],
    normalizedEntity,
  )) {
    const error = validateTraverseNodeForPublish(adminSchema, node);
    if (error) {
      errors.push(error);
    }
  }

  return errors;
}

describe('validateEntityInfo', () => {
  test('no type', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { type: '' } }),
      ),
    ).toMatchSnapshot();
  });

  test('invalid type', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { type: 'InvalidType' } }),
      ),
    ).toMatchSnapshot();
  });

  test('no authKey', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { authKey: null as unknown as string } }),
      ),
    ).toMatchSnapshot();
  });

  test('authKey not matching pattern', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { authKey: 'something else' } }),
      ),
    ).toMatchSnapshot();
  });

  test('no name', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { name: '' } }),
      ),
    ).toMatchSnapshot();
  });

  test('name with line break', () => {
    expect(
      validateEntityInfo(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_DEFAULT, { info: { name: 'Hello\nworld' } }),
      ),
    ).toMatchSnapshot();
  });
});

describe('validateEntityInfoForCreate', () => {
  test('no type', () => {
    expect(
      validateEntityInfoForCreate(adminSchema, ['entity'], {
        info: { type: '', name: 'No type' },
        fields: {},
      }),
    ).toMatchSnapshot();
  });

  test('invalid type', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { type: 'InvalidType' } }),
      ),
    ).toMatchSnapshot();
  });

  test('valid: no authKey (undefined)', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { authKey: undefined } }),
      ),
    ).toBeNull();
  });

  test('valid: no authKey (null)', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { authKey: null } }),
      ),
    ).toBeNull();
  });

  test('authKey not matching pattern', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { authKey: 'something else' } }),
      ),
    ).toMatchSnapshot();
  });

  test('no name', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { name: '' } }),
      ),
    ).toMatchSnapshot();
  });

  test('name with line break', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { name: 'Hello\nworld' } }),
      ),
    ).toMatchSnapshot();
  });

  test('invalid version', () => {
    expect(
      validateEntityInfoForCreate(
        adminSchema,
        ['entity'],
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { info: { version: 0 as 1 } }),
      ),
    ).toMatchSnapshot();
  });
});

describe('validateEntityInfoForUpdate', () => {
  test('change type', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { type: 'RichTextsEntity' }, fields: {} },
      ),
    ).toMatchSnapshot();
  });

  test('valid: same type', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { type: 'StringEntity' }, fields: {} },
      ),
    ).toBeNull();
  });

  test('change authKey', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { authKey: 'subject' }, fields: {} },
      ),
    ).toMatchSnapshot();
  });

  test('valid: no authKey (undefined)', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { authKey: undefined }, fields: {} },
      ),
    ).toBeNull();
  });

  test('valid: no authKey (null)', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { authKey: null }, fields: {} },
      ),
    ).toBeNull();
  });

  test('valid: same authKey', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: 'subject', version: 1 } },
        { id: '123', info: { authKey: 'subject' }, fields: {} },
      ),
    ).toBeNull();
  });

  test('name with line break', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { name: 'hello\nworld' }, fields: {} },
      ),
    ).toMatchSnapshot();
  });

  test('version with wrong value', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { version: 1 }, fields: {} },
      ),
    ).toMatchSnapshot();
  });

  test('valid: version with correct value', () => {
    expect(
      validateEntityInfoForUpdate(
        ['entity'],
        { info: { type: 'StringEntity', authKey: '', version: 1 } },
        { id: '123', info: { version: 2 }, fields: {} },
      ),
    ).toBeNull();
  });
});

describe('validateTraverseNodeForSave', () => {
  test('error', () => {
    expect(
      validateTraverseNodeForSave(adminSchema, {
        type: 'error',
        path: ['entity', 'foo'],
        errorType: ContentTraverseNodeErrorType.generic,
        message: 'Error message',
      }),
    ).toMatchSnapshot();
  });
});

describe('Validate entity shared', () => {
  test('Pass: required and adminOnly with no value', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { requiredAndAdminOnly: null } }),
      ),
    ).toEqual([]);
  });

  test('Fail: required with no value', () => {
    expect(
      validateEntity(copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { required: null } })),
    ).toMatchSnapshot();
  });

  test('Fail: extra entity fields', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { extra: 'extra', normal: 'normal' },
        }),
        { keepExtraFields: true },
      ),
    ).toMatchSnapshot();
  });
});

describe('Validate entity entity', () => {
  test('Fail: reference with extra keys', () => {
    expect(
      validateEntity(
        copyEntity(ENTITIES_ENTITY_DEFAULT, {
          fields: { normal: { id: '123', version: 123 } },
        }),
      ),
    ).toMatchSnapshot();
  });
});

describe('Validate entity location', () => {
  test('Fail: extra keys in location object', () => {
    expect(
      validateEntity(
        copyEntity(LOCATIONS_ENTITY_DEFAULT, {
          fields: { normal: { lat: 1, lng: 2, extra: 'extra' } },
        }),
      ),
    ).toMatchSnapshot();
  });
});

describe('Validate entity number', () => {
  test('Fail: integer with float value', () => {
    expect(
      validateEntity(copyEntity(NUMBERS_ENTITY_CREATE_DEFAULT, { fields: { integer: 1.2345 } })),
    ).toMatchSnapshot();
  });
});

describe('Validate entity richText', () => {
  test('Fail: rich text text node with line break', () => {
    expect(
      validateEntity(
        copyEntity(RICH_TEXTS_ENTITY_CREATE_DEFAULT, {
          fields: {
            anyNodes: createRichText([
              createRichTextParagraphNode([createRichTextTextNode('hello\nworld')]),
            ]),
          },
        }),
      ),
    ).toMatchSnapshot();
  });
});

describe('Validate entity string', () => {
  test('Pass: matchPattern matched string', () => {
    expect(
      validateEntity(copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { pattern: 'baz' } })),
    ).toEqual([]);
  });

  test('Fail: matchPattern unmatched string', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { pattern: 'will not match' } }),
      ),
    ).toMatchSnapshot();
  });

  test('Pass: matchPattern matched string items in list', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { patternList: ['foo', 'bar', 'baz'] },
        }),
      ),
    ).toEqual([]);
  });

  test('Fail: matchPattern unmatched string item in list', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { patternList: ['foo', 'will not match'] },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('Pass: multiline string', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { multiline: 'foo\nbar\nbaz' },
        }),
      ),
    ).toEqual([]);
  });

  test('Fail: multiline=false string with line break', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { normal: 'foo\nbar\nbaz' },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('Pass: values matched string', () => {
    expect(
      validateEntity(copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { values: 'baz' } })),
    ).toEqual([]);
  });

  test('Pass: values matched string list', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, {
          fields: { valuesList: ['foo', 'bar', 'baz'] },
        }),
      ),
    ).toEqual([]);
  });

  test('Fail: values unmatched string', () => {
    expect(
      validateEntity(copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { values: 'other' } })),
    ).toMatchSnapshot();
  });

  test('Fail: values unmatched string list', () => {
    expect(
      validateEntity(
        copyEntity(STRINGS_ENTITY_CREATE_DEFAULT, { fields: { valuesList: ['foo', 'other'] } }),
      ),
    ).toMatchSnapshot();
  });
});

describe('Validate entity component', () => {
  test('Fail: admin only component in normal field', () => {
    expect(
      validateEntity(
        copyEntity(VALUE_ITEMS_ENTITY_CREATE_DEFAULT, {
          fields: {
            any: { type: 'AdminOnlyComponent' },
          },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('Fail: wrong type, not matching componentTypes', () => {
    expect(
      validateEntity(
        copyEntity(VALUE_ITEMS_ENTITY_CREATE_DEFAULT, {
          fields: {
            stringsComponent: { type: 'NumbersComponent', integer: 1 },
          },
        }),
      ),
    ).toMatchSnapshot();
  });

  test('Fail: extra fields', () => {
    expect(
      validateEntity(
        copyEntity(VALUE_ITEMS_ENTITY_CREATE_DEFAULT, {
          fields: {
            any: { type: 'NumbersComponent', integer: 1, extra: 1.234 },
          },
        }),
        { keepExtraFields: true },
      ),
    ).toMatchSnapshot();
  });
});

describe('groupValidationIssuesByTopLevelPath', () => {
  test('no errors', () => {
    expect(groupValidationIssuesByTopLevelPath([])).toEqual({ root: [], children: new Map() });
  });

  test('root errors and list errors', () => {
    expect(
      groupValidationIssuesByTopLevelPath([
        { type: 'save', path: [], message: 'Root error' },
        { type: 'save', path: [0], message: 'Index 0 error' },
      ]),
    ).toEqual({
      root: [{ type: 'save', path: [], message: 'Root error' }],
      children: new Map([[0, [{ type: 'save', path: [], message: 'Index 0 error' }]]]),
    });
  });

  test('root errors and field errors', () => {
    expect(
      groupValidationIssuesByTopLevelPath([
        { type: 'save', path: [], message: 'Root error' },
        { type: 'save', path: ['field'], message: 'Field error' },
      ]),
    ).toEqual({
      root: [{ type: 'save', path: [], message: 'Root error' }],
      children: new Map([['field', [{ type: 'save', path: [], message: 'Field error' }]]]),
    });
  });
});
