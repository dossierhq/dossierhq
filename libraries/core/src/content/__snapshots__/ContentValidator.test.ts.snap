// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validate entity entity > Fail: reference with extra keys 1`] = `
[
  {
    "message": "Invalid keys for Entity: version",
    "path": [
      "entity",
      "fields",
      "normal",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity location > Fail: extra keys in location object 1`] = `
[
  {
    "message": "Invalid keys for Location: extra",
    "path": [
      "entity",
      "fields",
      "normal",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity number > Fail: integer with float value 1`] = `
[
  {
    "message": "Value must be an integer",
    "path": [
      "entity",
      "fields",
      "integer",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity richText > Fail: rich text text node with line break 1`] = `
[
  {
    "message": "Rich text text nodes cannot contain line breaks, use linebreak nodes instead",
    "path": [
      "entity",
      "fields",
      "anyNodes",
      0,
      0,
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity shared > Fail: extra entity fields 1`] = `
[
  {
    "message": "Invalid fields for entity of type StringsEntity: extra",
    "path": [
      "entity",
      "fields",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity shared > Fail: required with no value 1`] = `
[
  {
    "message": "Required field is empty",
    "path": [
      "entity",
      "fields",
      "required",
    ],
    "type": "publish",
  },
]
`;

exports[`Validate entity string > Fail: matchPattern unmatched string 1`] = `
[
  {
    "message": "Value does not match pattern fooBarBaz",
    "path": [
      "entity",
      "fields",
      "pattern",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity string > Fail: matchPattern unmatched string item in list 1`] = `
[
  {
    "message": "Value does not match pattern fooBarBaz",
    "path": [
      "entity",
      "fields",
      "patternList",
      1,
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity string > Fail: multiline=false string with line break 1`] = `
[
  {
    "message": "Value cannot contain line breaks",
    "path": [
      "entity",
      "fields",
      "normal",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity string > Fail: values unmatched string 1`] = `
[
  {
    "message": "Value does not match any of the allowed values",
    "path": [
      "entity",
      "fields",
      "values",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity string > Fail: values unmatched string list 1`] = `
[
  {
    "message": "Value does not match any of the allowed values",
    "path": [
      "entity",
      "fields",
      "valuesList",
      1,
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity valueItem > Fail: admin only value item in normal field 1`] = `
[
  {
    "message": "Value item of type AdminOnlyValueItem is adminOnly",
    "path": [
      "entity",
      "fields",
      "any",
    ],
    "type": "publish",
  },
]
`;

exports[`Validate entity valueItem > Fail: extra fields 1`] = `
[
  {
    "message": "Invalid fields for value item of type NumbersValueItem: extra",
    "path": [
      "entity",
      "fields",
      "any",
    ],
    "type": "save",
  },
]
`;

exports[`Validate entity valueItem > Fail: wrong type, not matching valueTypes 1`] = `
[
  {
    "message": "Value item of type NumbersValueItem is not allowed in field (supported types: StringsValueItem)",
    "path": [
      "entity",
      "fields",
      "stringsValueItem",
    ],
    "type": "save",
  },
]
`;

exports[`validateEntityInfo > authKey not matching pattern 1`] = `
{
  "message": "AuthKey 'something else' does not match pattern 'noneSubject' (^(none|subject)$)",
  "path": [
    "info",
    "authKey",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfo > invalid type 1`] = `
{
  "message": "Entity type InvalidType doesn’t exist",
  "path": [
    "entity",
    "info",
    "type",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfo > name with line break 1`] = `
{
  "message": "Name cannot contain line breaks",
  "path": [
    "entity",
    "info",
    "name",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfo > no authKey 1`] = `
{
  "message": "AuthKey is required",
  "path": [
    "entity",
    "info",
    "authKey",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfo > no name 1`] = `
{
  "message": "Name is required",
  "path": [
    "entity",
    "info",
    "name",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfo > no type 1`] = `
{
  "message": "Type is required",
  "path": [
    "entity",
    "info",
    "type",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > authKey not matching pattern 1`] = `
{
  "message": "AuthKey 'something else' does not match pattern 'noneSubject' (^(none|subject)$)",
  "path": [
    "info",
    "authKey",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > invalid type 1`] = `
{
  "message": "Entity type InvalidType doesn’t exist",
  "path": [
    "entity",
    "info",
    "type",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > invalid version 1`] = `
{
  "message": "Version must be 0 when creating a new entity",
  "path": [
    "entity",
    "info",
    "version",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > name with line break 1`] = `
{
  "message": "Name cannot contain line breaks",
  "path": [
    "entity",
    "info",
    "name",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > no authKey 1`] = `
{
  "message": "AuthKey is required",
  "path": [
    "entity",
    "info",
    "authKey",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > no name 1`] = `
{
  "message": "Name is required",
  "path": [
    "entity",
    "info",
    "name",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForCreate > no type 1`] = `
{
  "message": "Type is required",
  "path": [
    "entity",
    "info",
    "type",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForUpdate > change authKey 1`] = `
{
  "message": "New authKey subject doesn’t correspond to previous authKey none",
  "path": [
    "entity",
    "info",
    "authKey",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForUpdate > change type 1`] = `
{
  "message": "New type RichTextsEntity doesn’t correspond to previous type StringEntity",
  "path": [
    "entity",
    "info",
    "type",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForUpdate > name with line break 1`] = `
{
  "message": "Name cannot contain line breaks",
  "path": [
    "entity",
    "info",
    "name",
  ],
  "type": "save",
}
`;

exports[`validateEntityInfoForUpdate > version with wrong value 1`] = `
{
  "message": "The latest version of the entity is 1, so the new version must be 2 (got 1)",
  "path": [
    "entity",
    "info",
    "version",
  ],
  "type": "save",
}
`;

exports[`validateTraverseNodeForSave > error 1`] = `
{
  "message": "Error message",
  "path": [
    "entity",
    "foo",
  ],
  "type": "save",
}
`;
